### PR TITLE
chore(deps): Try LTO to see if it fixes linking issues on old OSes

### PR DESCRIPTION
**Pull request by jszwedko**
Seeing if this fixes `nix` compilation as suggested by https://github.com/nix-rust/nix/issues/1972#issuecomment-1521047819

Signed-off-by: Jesse Szwedko <jesse.szwedko@datadoghq.com>

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -35,11 +35,15 @@ path = "src/config/loading/secret_backend_example.rs"
 test = false
 bench = false
 
+[profile.dev]
+lto = true
```
Comment (bruceg) [inline Cargo.toml:39]: For this test, setting `lto = "thin"` is faster to link.

**Comment by jszwedko** [comment]
@zamazan4ik you might find this interesting too :) Turns out we weren't doing LTO on release assets.

**Comment by zamazan4ik** [comment]
Hah, I was sure it's enabled since Vector has https://github.com/vectordotdev/vector/blob/master/scripts/environment/release-flags.sh ... :)

**Comment by jszwedko** [comment]
> Hah, I was sure it's enabled since Vector has https://github.com/vectordotdev/vector/blob/master/scripts/environment/release-flags.sh ... :)

Aha, actually you are right :D We just don't use those flags for the performance tests apparently.

**Comment by jpds** [comment]
You might want to add the other optimization options from: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/Cargo.toml#L20-L24

Right now I'm seeing:

```
276M Jan 31 23:19 ../vector-non-lto*
146M Feb  7 18:42 target/release/vector* # with LTO
69M Feb  1 00:03 target/release/vector* # release profile options from c-h
```

**Comment by zamazan4ik** [comment]
How does enabling the `opt-level = "s"` option influence Vector performance? I think we cannot simply switch from `opt-level = 3` to `opt-level = "s"` without benchmarking. AFAIK, for Vector right now performance is more important than the binary size. And do not forget, the **actual** release flags for Vector are described here - https://github.com/vectordotdev/vector/blob/master/scripts/environment/release-flags.sh ;)

**Comment by jszwedko** [comment]
Vector releases do enable LTO (per https://github.com/vectordotdev/vector/pull/17342#issuecomment-1932674613 this is done by changing the flags in CI). The binaries we distribute are around ~120 MB. Certainly they could be smaller, but, as @zamazan4ik notes, Vector's focus is performance and so the compilation is optimized for that. I can see that Nix is sensitive to the package size though, maybe it'd make sense to use `opt-level = "s"` when compiling Vector for distribution on that platform?

**Comment by jszwedko** [comment]
I swear the binaries used to be smaller too, around 80 MB. I might bisect down to see if there were specific commits that bumped us up significantly or if it was a slow burn.

**Comment by jpds** [comment]
> Vector releases do enable LTO (per https://github.com/vectordotdev/vector/pull/17342#issuecomment-1932674613 this is done by changing the flags in CI). The binaries we distribute are around ~120 MB.

The issue is that no downstream Linux distribution is going to use the prebuilt binaries. I could patch this in, but much easier if this just gets merged.

I pushed [the same thing to another Rust project](https://git.deuxfleurs.fr/Deuxfleurs/garage/pulls/704) last month and it's even landed in the stable NixOS release as of yesterday:

```
.r-xr-xr-x 61M root  1 Jan  1970 /nix/store/98ka7zbb7x88vv23yic0h99nx1spv4s7-garage-0.9.0/bin/garage
.r-xr-xr-x 28M root  1 Jan  1970 /nix/store/dl68bfmrkbn98qlpa2i84j3qpxsixkzp-garage-0.9.2/bin/garage
```

**Comment by jszwedko** [comment]
> > Vector releases do enable LTO (per [#17342 (comment)](https://github.com/vectordotdev/vector/pull/17342#issuecomment-1932674613) this is done by changing the flags in CI). The binaries we distribute are around ~120 MB.
> 
> The issue is that no downstream Linux distribution is going to use the prebuilt binaries. I could patch this in, but much easier if this just gets merged.
> 
> I pushed [the same thing to another Rust project](https://git.deuxfleurs.fr/Deuxfleurs/garage/pulls/704) last month and it's even landed in the stable NixOS release as of yesterday:
> 
> ```
> .r-xr-xr-x 61M root  1 Jan  1970 /nix/store/98ka7zbb7x88vv23yic0h99nx1spv4s7-garage-0.9.0/bin/garage
> .r-xr-xr-x 28M root  1 Jan  1970 /nix/store/dl68bfmrkbn98qlpa2i84j3qpxsixkzp-garage-0.9.2/bin/garage
> ```

That's a good point :) I opened https://github.com/vectordotdev/vector/pull/20034. Let me know what you think of that.

**Comment by jszwedko** [comment]
I can't remember which OSes I was targeting with this change. I'm guessing it was CentOS 7, which went EOL earlier this year.

**Pull request closed without merging**